### ISSUE TITLE
feat: Support passing in DuckDBPyConnection into ibis.duckdb.connect()

### ISSUE THREAD

**Issue by NickCrews** [ibis-project/ibis#6636]
### Is your feature request related to a problem?

I am working with [splink](https://github.com/moj-analytical-services/splink). As part of their API you can get a hold of a duckdb.DuckDBPyConnection. I want to be able to query this existing connection.

### Describe the solution you'd like

I want to be able to make an ibis connection from this. This seems like a reasonable use case, where I want to connect to an existing DB connection instead of needing to create a new one. Could also make it so ibis.duckdb.connect accepts a sqlalchemy.Engine

### What version of ibis are you running?

5.1.0

### What backend(s) are you using, if any?

duckdb

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cpcloud** [comment]
Hey @NickCrews, thanks for the issue.

Is there a way you can flip the ownership of the connection around and pass the underlying connection from ibis to splink?

**Comment by NickCrews** [comment]
That actually is what I'm currently doing, but the only way I found was through a bunch of private stuff that I'm worried is fragile. If ibis exposed a more official API then that would work (for me at least)

**Comment by lostmygithubaccount** [comment]
how does one get the DuckDB connection from the Ibis connection? I need to do this as well, I thought it was `con.con` but that seems to be a SQLAlchemy object. is there an easy way to get the `duckdb.DuckDBPyConnection`?

**Comment by NickCrews** [comment]
```python
def ibis_backend_to_duckdb_connection(backend) -> duckdb.DuckDBPyConnection:
    sqlalchemy_con = backend.con
    duckdb_engine_wrapper: ConnectionWrapper = (
        sqlalchemy_con.raw_connection().dbapi_connection
    )
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        return duckdb_engine_wrapper.c
```

**Comment by cpcloud** [comment]
There's a lot of complexity in accessing the underlying connection due to the way SQLAlchemy works. That's not a value judgment about SQLAlchemy, it's just the way it works.

Maybe that's a good enough reason to provide an interface to the underlying connection now that we don't use connection pooling and there's only ever a single connection in play.

If so, the main thing to has out would be the interface:

Naming aside, what's the return type of the API? A DBAPI2 thing? Something else?

```
class Backend:
    ...

	def raw_con(self) -> ?:
		return ...
```

**Comment by NickCrews** [comment]
I'm not familiar at all with DBAPI. Can you convert both of those things to DuckdDBPyConnection? I think that is the lingua franca for any python lib that wants to use duckdb, so there needs to be a clean way to get there, even if that is a second step.

I'm also not familiar enough with all the various backends, but it seems like they are so varied that trying to come up with a uniform API is going to be futile. Like can you even make a DBAPI for the polars and pandas backends, which aren't SQL based?

**Comment by cpcloud** [comment]
> I'm also not familiar enough with all the various backends, but it seems like they are so varied that trying to come up with a uniform API is going to be futile. Like can you even make a DBAPI for the polars and pandas backends, which aren't SQL based?

I was trying to make a straw man argument, because I agree that DBAPI doesn't make sense for some backends.

I guess the real question is: should be there be any consistency between the output of the API, or is it effectively "you're stepping outside of ibis, you're on your own"?

**Comment by NickCrews** [comment]
> you're stepping outside of ibis

I think so yeah. As far as I know, If you have a DBAPI, a DuckDBPyConnection, google.cloud.bigquery.Cient(), or any other of the external representations, there is no way to plug them into Ibis. The starting step for interacting with Ibis is always making a Backend.

Ideally there would an inverse `ibis.from_raw_con(con)` (or `DuckDbBackend.from_raw_con(con)` to be explicit), and then that is your ticket to get back into ibis. But at least in my case, if I use/mutate the connection outside ibis (using PRAGMAs or something), the the connection inside ibis also sees the effect (and vice versa), so at least in my case I don't need both directions. But at least leaving the door open to this API in the future wouldn't be too hard?

**Comment by cpcloud** [comment]
As we move to sqlglot, the underlying `.con` will be the DBAPI. That should be sufficient to access the things you want.

**Comment by OlivierBinette** [comment]
Bumping this as I also work with separate tools that want a connection object like a PyDuckDBConnection. Right now, I have to keep two separate connections to my DuckDB database: one for data manipulation with Ibis, and another for my other tools.

Having two separate connections creates problems as they do not share temp tables. When working with in-memory databases, the two connections will refer to two different in-memory databases, which is not ideal.

It would also be OK if I can recover some connection object from Ibis. That approach makes sense if we use Ibis as stable interface to a backend that could change. But if we only use Ibis as a high-level convenience tool, I would prefer having Ibis depend on my connection object.

**Comment by cpcloud** [comment]
> sounds good! Will that be in 8.0.0? Any timeline on when we can expect that?

Hard to say, but we're targeting 9.0.0 for the sqlglot refactors (they are quite large). I would expect 9.0.0 probably middle to end of Q1 2024 depending on how things go.

Once we get to a point where we have CI passing we'd love to get some beta testing going with you and whomever else is interested to shake out any remaining bugs.

Going forward we are going to try to ensure that our SQL backends use only DBAPIs not wrapped by any SQLAlchemy layers.

**Comment by cpcloud** [comment]
@OlivierBinette Does @NickCrews's workaround above work for you for pulling the connection out of an Ibis backend?

**Comment by OlivierBinette** [comment]
@cpcloud I think NickCrew's solution would work, but I prefer to keep two separate connections to the same object as it is ok for my problem and is more robust to changes to Ibis.

**Comment by NickCrews** [comment]
followup: with the sqlglot refactor, on ibis main you can now just get this directly! `ddb_conn: duckdb.DuckDBPyConnection = backend.con` 🥳 @cpcloud can we rely on this as a public API?

**Comment by NickCrews** [comment]
gotcha, thanks!

are you thinking that all backends have a `con` attribute, but each is backend-specific? eg `SqliteBackend.con` is a `sqlite3.Connection`?

**Comment by cpcloud** [comment]
Yeah, we're not getting into the business of wrapping DB API things.

**Comment by cpcloud** [comment]
Even if something isn't a DBAPI, and it's called `con` we're not going to wrap it. The whole point of this IIUC is to have something that is backend specific.

**Comment by NickCrews** [comment]
> The whole point of this IIUC is to have something that is backend specific.

Yes exactly, this sounds like what we want. Thanks for the clarifications, `backend.con` is exactly what I and @OlivierBinette needed
